### PR TITLE
modification to `config.option()` and `config.merge()` to support

### PR DESCRIPTION
**Pull request by gladiatr72**
### What does this PR do?

Modifies the mechanism used to resolve dictionary key paths.  Rather than using a single level dictionary-lookup, this uses salt.utils.traverse_dict()
### What issues does this PR fix or reference?

I am submitting this as a precursor to fixing the boto execution modules disfunction with regards to auth profiles

https://github.com/saltstack/salt/pull/30158
### Tests written?

No (sorry--time 😞 )

multi-dimensional pillar support.

The existing implementation only takes care of single-depth pillar keys.
This patch maintains the existing functionality but rather than use
simple single-key dictionary lookups, this uses
`salt.utils.traverse_dict()` calls.

e.g.:
`salt-call config.option cloud.config.ec2` will properly return a value
from **opts**, **pillar**.master (if not disabled) or  **pillar**

similarly, config.merge() has also been updated to support the same
lookups while maintaining current merge logic.

### CODE REVIEW

**Comment by cachedout** [comment]
We'll need to look very carefully here, since this has the potential to break many things if we aren't very careful. I might need a few days to fully review this since it's the week of SaltConf. Thanks in advance for your patience on this one. :]

**Comment by gladiatr72** [comment]
No worries at all.  I actually put a log line in my local copy (shouts, "CONFIG_OPTION/MERGE") to watch for any exceptions surrounding calls to those functions. Also no problem with regards to the schedule.  I hope things are going well at the conference! 😀

**Comment by gladiatr72** [comment]
FYI: I've been running my production and dev systems with this for the last 2.5 weeks without incident.

**Comment by cachedout** [comment]
@thatch45 and @s0undt3ch could you both look at this? I'm fairly sure this is benign but would like to see some more eyes on it.

**Comment by thatch45** [comment]
Is there any good reason to use a `.` instead of the standard `:`? I think this will be most confusing. Other than that I think this should be ok

**Comment by gladiatr72** [comment]
@thatch I think the idea was to keep the `config.option()` interface uniform so the requested object could be delivered either through the server config or pillar or whatnot.  Are you thinking about a delimiter argument for `config.(option|get)()`?

File: salt/modules/config.py [hunk 1]
```diff
@@ -136,23 +137,43 @@ def option(
...
+
     '''
+
+    log.debug('CONFIG.OPTION: %s', value)
     if not omit_opts:
-        if value in __opts__:
-            return __opts__[value]
+        retval = traverse_dict(__opts__, value, None, delimiter='.')
+        if retval:
+            log.debug('CONFIG.OPTION (opts): %s', str(retval))
```
Comment (s0undt3ch) [inline salt/modules/config.py]: No need to call `str()`, using string substitution will do it for you.

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by cachedout** [comment]
I know this is really late but now that we have the carbon release out the door I would like to come back and try to get this into Nitrogen. @gladiatr72 any chance you could rebase this? If so, I'd like to get it merged. :] Apologies for the absurd delay on this.

**Comment by cachedout** [comment]
Closed for no reply. We'd love to get this in but we need a response. Thanks.

**Pull request closed without merging**